### PR TITLE
Change in PackedByteArray decode api and docs

### PR DESCRIPTION
**Pull request by ChristopheClaustre**
- New methods PackedByteArray::to_***_array to decode all the byte array at once (and be easier to use) and its documentations
- Documentation for Packed***Array for the counterpart method to_byte_array

Discussed here : https://github.com/godotengine/godot-proposals/issues/2993

### CODE REVIEW

**Comment by pycbouh** [comment]
You are missing empty lines between method definitions (and you've removed some that existed). This will fail our CI check for code style, so it needs to be fixed.

Consider installing pre-commit hooks too to perform some of those checks locally before pushing.

**Comment by ChristopheClaustre** [comment]
@pycbouh Currently installing the hooks ;)
Also I see that linux build doesn't work... I understand how to fix but memcpy is available in the global scope directly ?

First pull request with github here, I suppose it works the way I am used to.
If I commit on my fork on the same branch it will update the PR here ?

**Comment by pycbouh** [comment]
> If I commit on my fork on the same branch it will update the PR here ?

Yes, but make sure to rebase and squash your commits. We usually accept only one commit per PR unless there is a good reason to split them.

> Also I see that linux build doesn't work... I understand how to fix but memcpy is available in the global scope directly ?

Sorry, can't help you with that, but someone else might. What I can say is that we are very conservative with the usage of STL, but `memcpy` seems to be used directly wherever it is used, without `std::`. Can be some implicit include from another include that does that.

PS. Strangely, your lack of empty lines didn't trigger the clang-format on CI, but that's definitely not following our codestyle, so still needs to be amended.

**Comment by ChristopheClaustre** [comment]
> Yes, but make sure to rebase and squash your commits.

I am used to auto squash on pull request so I didn't do it.
But since my commit are on the remote I need to force push on my remote to do that now ?

> Sorry, can't help you with that, but someone else might.

As clang is a good guy usually, I will believe in what he says and will remove the `std::`.

**Comment by pycbouh** [comment]
> But since my commit are on the remote I need to force push on my remote to do that now ?

Yes, whatever you do to your branch in your fork/remote it will update this PR. GitHub is even being ever-helpful and reminding everyone about that here: 🙃 

image

**Comment by ChristopheClaustre** [comment]
Github is nice :)
Thank you a lot for your help @pycbouh
Do you know how I can relaunch the CI ? To know if my commits are now resolving the problem ?
Then I will squash them all.

**Comment by pycbouh** [comment]
As you are a first-time contributor a member needs to approve each run. I approved it. Once this is merged, your next PRs will run automatically.

**Comment by ChristopheClaustre** [comment]
@pycbouh so i am really sorry to bother you with my trivial PR ^^"
I will need one more check, then squash and the PR is okay unless somebody has something to say on it

**Comment by ChristopheClaustre** [comment]
@pycbouh Unless somebody as something to say, my PR is ready to be merged
Should I do something on my side ?

**Comment by pycbouh** [comment]
@ChristopheClaustre No, just wait for reviews. Feel free to drop by the contributors chat if you want to discuss it with someone life. I personally cannot review this, so I can't help with that. But the appropriate teams are assigned, so you probably just need to wait a bit.

**Comment by clayjohn** [comment]
I spoke to reduz on the contributor chat and he confirmed that he supports adding the functions!

We just need someone who is familiar enough with this stuff to approve the implementation. Perhaps @vnen, @akien-mga, or @JFonS ?

File: doc/classes/PackedByteArray.xml
Comment (akien-mga) [inline doc/classes/PackedByteArray.xml:524]:
```suggestion
				Returns a copy of the data converted to a [PackedFloat32Array], where each block of 4 bytes has been converted to a 32-bit float (C++ [code]float[/code]).
				The size of the new array will be [code]byte_array.size() / 4[/code].
				If the original data can't be converted to 32-bit floats, the resulting data is undefined.
```

File: doc/classes/PackedByteArray.xml
Comment (akien-mga) [inline doc/classes/PackedByteArray.xml:533]:
```suggestion
				Returns a copy of the data converted to a [PackedFloat64Array], where each block of 8 bytes has been converted to a 64-bit float (C++ [code]double[/code], Godot [float]).
				The size of the new array will be [code]byte_array.size() / 8[/code].
				If the original data can't be converted to 64-bit floats, the resulting data is undefined.
```

File: doc/classes/PackedByteArray.xml
Comment (akien-mga) [inline doc/classes/PackedByteArray.xml:542]:
```suggestion
				Returns a copy of the data converted to a [PackedInt32Array], where each block of 4 bytes has been converted to a signed 32-bit integer (C++ [code]int32_t[/code]).
				The size of the new array will be [code]byte_array.size() / 4[/code].
				If the original data can't be converted to signed 32-bit integers, the resulting data is undefined.
```

File: doc/classes/PackedByteArray.xml
Comment (akien-mga) [inline doc/classes/PackedByteArray.xml:551]:
```suggestion
				Returns a copy of the data converted to a [PackedInt64Array], where each block of 4 bytes has been converted to a signed 64-bit integer (C++ [code]int64_t[/code], Godot [int]).
				The size of the new array will be [code]byte_array.size() / 8[/code].
				If the original data can't be converted to signed 64-bit integers, the resulting data is undefined.
```

File: doc/classes/PackedFloat32Array.xml
Comment (akien-mga) [inline doc/classes/PackedFloat32Array.xml:205]:
```suggestion
				Returns a copy of the data converted to a [PackedByteArray], where each element have been encoded as 4 bytes.
				The size of the new array will be [code]float32_array.size() * 4[/code].
```

File: doc/classes/PackedFloat64Array.xml
Comment (akien-mga) [inline doc/classes/PackedFloat64Array.xml:205]:
```suggestion
				Returns a copy of the data converted to a [PackedByteArray], where each element have been encoded as 8 bytes.
				The size of the new array will be [code]float64_array.size() * 8[/code].
```

File: doc/classes/PackedInt32Array.xml
Comment (akien-mga) [inline doc/classes/PackedInt32Array.xml:205]:
```suggestion
				Returns a copy of the data converted to a [PackedByteArray], where each element have been encoded as 4 bytes.
				The size of the new array will be [code]int32_array.size() * 4[/code].
```

File: doc/classes/PackedInt64Array.xml
Comment (akien-mga) [inline doc/classes/PackedInt64Array.xml:205]:
```suggestion
				Returns a copy of the data converted to a [PackedByteArray], where each element have been encoded as 8 bytes.
				The size of the new array will be [code]int64_array.size() * 8[/code].
```

**Review by akien-mga — APPROVED** [review]
The added code looks good to me. I made some suggestions to improve grammar in the docs, but overall it's fine :+1:

**Review by JFonS — APPROVED** [review]
I can't comment much on the bindings, but conversion code looks good :)

**Comment by ChristopheClaustre** [comment]
> The added code looks good to me. I made some suggestions to improve grammar in the docs, but overall it's fine 👍

@akien-mga I committed your suggestions. We just need to wait for the workflows to succeed

**Comment by akien-mga** [comment]
Looks good. Could you squash the two commits together?

**Comment by ChristopheClaustre** [comment]
@akien-mga  Done !

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**
